File: node/src/api/arch/role.js
/**
 * Copyright (c) 2015-2022, FineX, All Rights Reserved.
 * @author Thomas.Zhuang
 * @date 2022/06/08
 */

import { ApiClient } from '../../base/api_client';
import { ROLE_BASE_PATH } from '../../constants/http_constant';

export class RoleApiClient extends ApiClient {
    validVersions = ['v5'];
    defaultVersion = 'v5';


    /**
     * check version
     */
    async doRequest(options) {
        if (!this.validVersions.includes(this.version)) {
            this.version = this.defaultVersion;
        }
        return super.doRequest(options);
    }

    /**
     * 列出角色
     */
    async roleList(options) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'list',
            payload: {
                ...options
            }
        });
    }

    /**
     * 创建角色
     */
    async roleCreate(name, groupNo) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'create',
            payload: {
                name,
                group_no: groupNo
            }
        });
    }

    /**
     * 更新角色
     */
    async roleUpdate(roleNo, groupNo, options) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'update',
            payload: {
                role_no: roleNo,
                group_no: groupNo,
                name: options.name
            }
        });
    }

    /**
     * 删除角色
     */
    async roleDelete(roleNo) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'delete',
            payload: {
                role_no: roleNo
            }
        });
    }

    /**
     * 列出角色下所有成员
     */
    async roleMemberList(roleNo, options) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'user/list',
            payload: {
                role_no: roleNo,
                ...options
            }
        });
    }

    /**
     * 为自建角色批量添加成员
     */
    async roleAddMembers(roleNo, usernames) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'add_members',
            payload: {
                role_no: roleNo,
                usernames
            }
        });
    }

    /**
     * 为自建角色批量移除成员
     */
    async roleRemoveMembers(roleNo, usernames) {
        return await this.doRequest({
            method: 'POST',
            path: ROLE_BASE_PATH + 'remove_members',
            payload: {
                role_no: roleNo,
                usernames
            }
        });
    }
}

File: csharp/src/api/arch/Role.cs
/**
* Copyright (c) 2015-2022, FineX, All Rights Reserved.
* @author Thomas.Zhuang
* @date 2022/06/14
*/
using System.Text.Json;
using Src.Base;

namespace Src.Api.Jdy;

public class RoleApiClient : ApiClient
{
    private readonly static string[] validVersions = { "v2" };
    private readonly static string defaultVersion = "v2";
    public RoleApiClient(string host, string apiKey, string version = "") : base(host, apiKey, version)
    {
    }

    public new async Task<JsonElement?> doRequest(IDictionary<string, object> option)
    {
        if (Array.IndexOf(validVersions, base.version) == -1)
        {
            base.version = defaultVersion;
        }
        return await base.doRequest(option);
    }

    /**
    * 角色列表
    */
    public async Task<JsonElement?> roleList(IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/list");
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 角色创建
    */
    public async Task<JsonElement?> roleCreate(string name, int groupNo)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/create");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("name", name);
        payload.Add("group_no", groupNo);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }

    /**
    * 角色更新
    */
    public async Task<JsonElement?> roleUpdate(int roleNo, int groupNo, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/update");
        option.Add("role_no", roleNo);
        option.Add("group_no", groupNo);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 角色删除
    */
    public async Task<JsonElement?> roleDelete(int roleNo)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/delete");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("role_no", roleNo);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }

    /**
    * 列出角色下所有成员
    */
    public async Task<JsonElement?> roleMemberList(int roleNo, IDictionary<string, object> option)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/member_list");
        option.Add("role_no", roleNo);
        requestOption.Add("payload", option);
        return await this.doRequest(requestOption);
    }

    /**
    * 为自建角色批量添加成员
    */
    public async Task<JsonElement?> roleAddMembers(int roleNo, string[] usernames)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/add_members");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("role_no", roleNo);
        payload.Add("usernames", usernames);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }

    /**
    * 为自建角色批量移除成员
    */
    public async Task<JsonElement?> roleRmoveMembers(int roleNo, string[] usernames)
    {
        IDictionary<string, object> requestOption = new Dictionary<string, object>();
        requestOption.Add("method", "POST");
        requestOption.Add("path", "role/remove_members");
        IDictionary<string, object> payload = new Dictionary<string, object>();
        payload.Add("role_no", roleNo);
        payload.Add("usernames", usernames);
        requestOption.Add("payload", payload);
        return await this.doRequest(requestOption);
    }
}